File: android/lib/src/main/java/com/swmansion/gesturehandler/PinchGestureHandler.java
package com.swmansion.gesturehandler;

import android.content.Context;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.view.ViewConfiguration;

public class PinchGestureHandler extends GestureHandler<PinchGestureHandler> {

  private ScaleGestureDetector mScaleGestureDetector;
  private double mLastScaleFactor;
  private double mLastVelocity;

  private float mStartingSpan;
  private float mSpanSlop;

  private float focalX = Float.NaN;
  private float focalY = Float.NaN;

  private ScaleGestureDetector.OnScaleGestureListener mGestureListener =
          new ScaleGestureDetector.OnScaleGestureListener() {

    @Override
    public boolean onScale(ScaleGestureDetector detector) {
      double prevScaleFactor = mLastScaleFactor;
      mLastScaleFactor *= detector.getScaleFactor();
      long delta = detector.getTimeDelta();
      if (delta > 0) {
        mLastVelocity = (mLastScaleFactor - prevScaleFactor) / delta;
      }
      if (Math.abs(mStartingSpan - detector.getCurrentSpan()) >= mSpanSlop
              && getState() == STATE_BEGAN) {
        activate();
      }
      return true;
    }

    @Override
    public boolean onScaleBegin(ScaleGestureDetector detector) {
      mStartingSpan = detector.getCurrentSpan();
      return true;
    }

    @Override
    public void onScaleEnd(ScaleGestureDetector detector) {
      // ScaleGestureDetector thinks that when fingers are 27mm away that's a sufficiently good
      // reason to trigger this method giving us no other choice but to ignore it completely.
    }
  };

  public PinchGestureHandler() {
    setShouldCancelWhenOutside(false);
  }

  @Override
  protected void onHandle(MotionEvent event) {
    int state = getState();
    if (state == STATE_UNDETERMINED) {
      Context context = getView().getContext();
      mLastVelocity = 0f;
      mLastScaleFactor = 1f;
      mScaleGestureDetector = new ScaleGestureDetector(context, mGestureListener);
      ViewConfiguration configuration = ViewConfiguration.get(context);
      mSpanSlop = configuration.getScaledTouchSlop();
    }

    int activePointers = event.getPointerCount();
    if (event.getActionMasked() == MotionEvent.ACTION_POINTER_UP) {
      activePointers -= 1;
    }

    if (
      mScaleGestureDetector != null &&
      (state != STATE_ACTIVE || activePointers >= 2)
    ) {
      this.focalX = mScaleGestureDetector.getFocusX();
      this.focalY = mScaleGestureDetector.getFocusY();
    }

    if (state == STATE_UNDETERMINED) {
      begin();
    }

    if (mScaleGestureDetector != null) {
      mScaleGestureDetector.onTouchEvent(event);
    }

    if (state == STATE_ACTIVE && activePointers < 2) {
      end();
    } else if (event.getActionMasked() == MotionEvent.ACTION_UP) {
      fail();
    }
  }

  @Override
  protected void onReset() {
    mScaleGestureDetector = null;
    mLastVelocity = 0f;
    mLastScaleFactor = 1f;
  }

  public double getScale() {
    return mLastScaleFactor;
  }

  public double getVelocity() {
    return mLastVelocity;
  }

  public float getFocalPointX() {
    return this.focalX;
  }

  public float getFocalPointY() {
    return this.focalY;
  }
}

File: android/lib/src/main/java/com/swmansion/gesturehandler/PinchGestureHandler.kt
package com.swmansion.gesturehandler

import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.ScaleGestureDetector.OnScaleGestureListener
import android.view.ViewConfiguration
import kotlin.math.abs

class PinchGestureHandler : GestureHandler<PinchGestureHandler>() {
  var endOnFingerRelease = false

  var scale = 0.0
    private set
  var velocity = 0.0
    private set
  val focalPointX: Float
    get() = scaleGestureDetector?.focusX ?: Float.NaN
  val focalPointY: Float
    get() = scaleGestureDetector?.focusY ?: Float.NaN

  private var scaleGestureDetector: ScaleGestureDetector? = null
  private var startingSpan = 0f
  private var spanSlop = 0f
  private val gestureListener: OnScaleGestureListener = object : OnScaleGestureListener {
    override fun onScale(detector: ScaleGestureDetector): Boolean {
      val prevScaleFactor: Double = scale
      scale *= detector.scaleFactor.toDouble()
      val delta = detector.timeDelta
      if (delta > 0) {
        velocity = (scale - prevScaleFactor) / delta
      }
      if (abs(startingSpan - detector.currentSpan) >= spanSlop
        && state == STATE_BEGAN) {
        activate()
      }
      return true
    }

    init {
      setShouldCancelWhenOutside(false)
    }

    override fun onScaleBegin(detector: ScaleGestureDetector): Boolean {
      startingSpan = detector.currentSpan
      return true
    }

    override fun onScaleEnd(detector: ScaleGestureDetector) {
      // ScaleGestureDetector thinks that when fingers are 27mm away that's a sufficiently good
      // reason to trigger this method giving us no other choice but to ignore it completely.
    }
  }

  override fun onHandle(event: MotionEvent) {
    if (state == STATE_UNDETERMINED && event.pointerCount >= 2) {
      val context = view!!.context
      velocity = 0.0
      scale = 1.0
      scaleGestureDetector = ScaleGestureDetector(context, gestureListener)
      val configuration = ViewConfiguration.get(context)
      spanSlop = configuration.scaledTouchSlop.toFloat()
      begin()
    }
    if (state == STATE_BEGAN || state == STATE_ACTIVE) {
      scaleGestureDetector?.onTouchEvent(event)
    }
    var activePointers = event.pointerCount
    if (event.actionMasked == MotionEvent.ACTION_POINTER_UP || event.action == MotionEvent.ACTION_UP) {
      activePointers -= 1
    }

    if (state == STATE_ACTIVE) {
      if ((endOnFingerRelease && activePointers < 2) || activePointers == 0) {
        end()
      }
    } else if (event.actionMasked == MotionEvent.ACTION_UP) {
      fail()
    }
  }

  override fun needsToPreventOtherFromActivating(handler: GestureHandler<*>): Boolean {
    if (handler is NativeViewGestureHandler) {
      return true
    }

    return super.needsToPreventOtherFromActivating(handler)
  }

  override fun onReset() {
    scaleGestureDetector = null
    velocity = 0.0
    scale = 1.0
  }
}
